pkg/graph/builder.go

```diff
@@ -714,155 +674,192 @@ func extractDependencies(env *cel.Env, expression string, resourceNames []string
 	return dependencies, isStatic, nil
 }
 
-// validateResourceCELExpressions tries to validate the CEL expressions in the
-// resources against the resources defined in the resource graph definition.
-//
-// In this process, we pin a resource and evaluate the CEL expressions in the
-// context of emulated resources. Meaning that given 3 resources A, B, and C,
-// we evaluate A's CEL expressions against 2 emulated resources B and C. Then
-// we evaluate B's CEL expressions against 2 emulated resources A and C, and so
-// on.
-func validateResourceCELExpressions(resources map[string]*Resource, instance *Resource) error {
-	resourceIDs := maps.Keys(resources)
-	// We also want to allow users to refer to the instance spec in their expressions.
-	resourceIDs = append(resourceIDs, "schema")
-
-	env, err := krocel.DefaultEnvironment(krocel.WithResourceIDs(resourceIDs))
-	if err != nil {
-		return fmt.Errorf("failed to create CEL environment: %w", err)
-	}
-	instanceEmulatedCopy := instance.emulatedObject.DeepCopy()
-	if instanceEmulatedCopy != nil && instanceEmulatedCopy.Object != nil {
-		delete(instanceEmulatedCopy.Object, "apiVersion")
-		delete(instanceEmulatedCopy.Object, "kind")
-		delete(instanceEmulatedCopy.Object, "status")
-	}
-
-	// create includeWhenContext
-	includeWhenContext := map[string]*Resource{}
-	// For now, we will only support the instance context for includeWhen expressions.
-	// With this decision, we will decide on creation time and update time
-	// If we'll be creating resources or not
-	includeWhenContext["schema"] = &Resource{
-		emulatedObject: &unstructured.Unstructured{
-			Object: instanceEmulatedCopy.Object,
-		},
-	}
-
-	// create expressionsContext
-	expressionContext := map[string]*Resource{}
-	// add instance spec to the context
-	expressionContext["schema"] = &Resource{
-		emulatedObject: &unstructured.Unstructured{
-			Object: instanceEmulatedCopy.Object,
-		},
-	}
-	// include all resources, and remove individual ones
-	// during the validation
-	// this is done to avoid having to create a new context for each resource
-	for resourceName, contextResource := range resources {
-		expressionContext[resourceName] = contextResource
+// validateNode validates all CEL expressions for a single resource node:
+// - Template expressions (resource field values)
+// - includeWhen expressions (conditional resource creation)
+// - readyWhen expressions (resource readiness conditions)
+func validateNode(resource *Resource, templatesEnv, schemaEnv *cel.Env, resourceSchema *spec.Schema) error {
```

Comment:
Completely optional but something I noticed. 

Each of these validations are derivations of the data `resource` and `schema`. In the function outside `validateNode`, we scope down `templatesEnv` and `schemaEnv` and then inside this function, we scope down the `resourceEnv` for the readyWhen validation. 

It made me wonder two things:
1. If this pattern/separation was intentional (i.e., we create a schemaEnv outside, even if includeWhenExpressions == 0, but we check readyWhenExpressions exist before constructing the limited env)
2. If we ever might want to expand the language to support cross-resource readyWhens or status-based includeWhens. The latter is maybe more likely?